rfcs/0002-assetid.md

```diff
@@ -0,0 +1,107 @@
+This allows the Builder to be a pure function which makes it a lot easier to implement things like build caching and distributed building.
+
+### Load
+[load]: #load
+All loading will reference assets using AssetID. Different Resolver implementations will handle resolving an AssetID to a loadable asset blob depending on the runtime environment and the AssetID variant. The Resolver replaces the existing Source trait as a more general solution.
+
+For example, in a development environment the Resolver will probably connect to AssetHub to resolve the AssetID to a build artifact hash based on the target platform, then load the build artifact and return it to the caller.
+
+In a distributable build with optimized packed of assets, the Resolver may have an in-memory map of AssetUUID->FileLocation for knowing where an asset can be loaded from. It then performs the file read and returns the build artifact. For FilePath variants, the Resolver may have an in-memory map of FilePath->AssetUUID it uses before running the AssetUUID resolving process. 
+In the case of modding or asset overrides, a completely custom Resolver may be written that maintains a map of AssetUUID->FilePath to see if the user has provided a file override. If not, the resolver fallbacks to the resolver for packed assets.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The drawbacks compared to no change at all are possibly increased complexity?
+
+# Rationale, Alternatives and Prior Art
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design for an asset identifier is primarily inspired by the Unity asset management system and most of the advantages can be seen by looking at how their builds work. The most important improvement upon their design being the following:
+- Unity defines a UUID to reference a source file, then uses a "file-internal ID" to reference specific assets. This proposal improves upon that design with a general design that enables referencing specific assets within a source file with 16 bytes instead of 16 + 4 bytes.
+[Reference here](https://docs.unity3d.com/ScriptReference/Build.Content.ObjectIdentifier.html)
+
+The design is in my opinion superior to Godot, Unreal and Crytek asset management systems as their asset references use paths, leading to issues with renaming or moving files. Unreal's solution to moving/renaming only works if you move the file in the editor: put a "Redirector" at the path of a moved or renamed file. [Source here](https://docs.unrealengine.com/en-us/Engine/Basics/Redirectors)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+An unresolved question I have is whether URI should be included at all (not needed for most common use case, but has been requested previously) and if it is to be included, which URI library should be used.
```

Comment:
if we leave the special cases of paths or uris out of the system, this solves itself.